ros/lib/kessel/kessel_shared.py

```diff
@@ -0,0 +1,99 @@
+import grpc
+from ros.lib.config import get_logger
+from kessel.inventory.v1beta2 import inventory_service_pb2_grpc
+
+LOG = get_logger(__name__)
+
+# Singleton storage
+_GRPC_CHANNEL = None
+_GRPC_STUB = None
+_AUTH_CREDENTIALS = None
+
+
+def get_cached_kessel_auth_credentials():
+    """
+    Get or create cached OAuth2 credentials to avoid expensive OIDC discovery calls.
+    This is a shared cache used by both the gRPC singleton and KesselClient instances.
+    Returns:
+        OAuth2ClientCredentials object or None if creation fails
+    """
+    global _AUTH_CREDENTIALS
+    if _AUTH_CREDENTIALS is None:
+        from ros.lib.config import create_kessel_oauth2_credentials
+        LOG.debug("Creating shared OAuth2 credentials cache (first time)")
+        _AUTH_CREDENTIALS = create_kessel_oauth2_credentials()
+    return _AUTH_CREDENTIALS
+
+
+def get_kessel_stub(host):
+    """
+    Get or create a Kessel gRPC stub with OAuth2 authentication.
+
+    Args:
+        host: Kessel service host and port
+
+    Returns:
+        KesselInventoryServiceStub instance
+    """
+    global _GRPC_CHANNEL, _GRPC_STUB
+
+    if _GRPC_STUB is None:
+        try:
+            from ros.lib.config import (
+                KESSEL_AUTH_CLIENT_ID,
+                KESSEL_AUTH_CLIENT_SECRET,
+                KESSEL_AUTH_OIDC_ISSUER,
+                KESSEL_INSECURE
+            )
+
+            if KESSEL_AUTH_CLIENT_ID and KESSEL_AUTH_CLIENT_SECRET and KESSEL_AUTH_OIDC_ISSUER:
```

Comment:
These are not strictly required in this step, although it's true for calling RBAC. 
I would remove the check for these 3 configs from here.